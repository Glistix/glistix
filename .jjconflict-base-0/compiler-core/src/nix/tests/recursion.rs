use crate::assert_nix;

#[test]
fn tco() {
    assert_nix!(
        r#"
pub fn main(x) {
  case x {
    0 -> Nil
    _ -> main(x - 1)
  }
}
"#
    );
}

#[test]
fn tco_case_block() {
    assert_nix!(
        r#"
pub fn main(x) {
  case x {
    0 -> Nil
    _ -> {
      let y = x
      main(y - 1)
    }
  }
}
"#
    );
}

// https://github.com/gleam-lang/gleam/issues/1779
#[test]
fn not_tco_due_to_assignment() {
    assert_nix!(
        r#"
pub fn main(x) {
  let z = {
    let y = x
    main(y - 1)
  }
  z
}
"#
    );
}

// https://github.com/gleam-lang/gleam/issues/2400
#[test]
fn shadowing_so_not_recursive() {
    // This funtion is calling an argument with the same name as itself, so it is not recursive
    assert_nix!(
        r#"
pub fn map(map) {
  map()
}
"#
    );
}
