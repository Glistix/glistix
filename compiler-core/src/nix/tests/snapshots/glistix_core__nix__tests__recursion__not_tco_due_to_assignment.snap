---
source: compiler-core/src/nix/tests/recursion.rs
expression: "\npub fn main(x) {\n  let z = {\n    let y = x\n    main(y - 1)\n  }\n  z\n}\n"
snapshot_kind: text
---
----- SOURCE CODE

pub fn main(x) {
  let z = {
    let y = x
    main(y - 1)
  }
  z
}


----- COMPILED NIX
let main = x: let z = let y = x; in main (y - 1); in z; in { inherit main; }
